### PR TITLE
OCPBUGS-29435: Use structured authentication config

### PR DESCRIPTION
**Pull request by sjenning**
supercedes https://github.com/openshift/hypershift/pull/3573

requires https://github.com/openshift/kubernetes/pull/1881

bump to o/api to get this is also required https://github.com/openshift/api/pull/1771

### CODE REVIEW

File: control-plane-operator/controllers/hostedcontrolplane/kas/auth.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	serializedConfig, err := json.Marshal(authConfig)
+	if err != nil {
+		return fmt.Errorf("failed to serialize kube apiserver authentication config: %w", err)
+	}
+	config.Data[AuthenticationConfigKey] = string(serializedConfig)
+	return nil
+}
+
+func generateAuthConfig(spec *configv1.AuthenticationSpec) *apiserverv1alpha1.AuthenticationConfiguration {
+	config := &apiserverv1alpha1.AuthenticationConfiguration{
```
Comment (deads2k) [inline control-plane-operator/controllers/hostedcontrolplane/kas/auth.go:31]: to backport to 4.15, we'll need a copy of this API type for rendering.

File: control-plane-operator/controllers/hostedcontrolplane/kas/config.go [hunk 2]
```diff
@@ -173,25 +174,7 @@ func generateConfig(p KubeAPIServerConfigParams, version semver.Version) *kcpv1.
...
-			args.Set("oidc-groups-prefix", provider.ClaimMappings.Groups.Prefix)
-			for _, cvr := range provider.ClaimValidationRules {
-				// TODO: currently can only support a single required claim because of how kubeAPIServerArgs dedups config fields
-				// In order to specify multiple required claims, the flag must be used multiple times
-				// https://kubernetes.io/docs/reference/access-authn-authz/authentication/#configuring-the-api-server
-				args.Set("oidc-required-claim", fmt.Sprintf("%s=%s", cvr.RequiredClaim.Claim, cvr.RequiredClaim.RequiredValue))
-			}
-			args.Set("oidc-ca-file", oidcCAFile(provider.Issuer.CertificateAuthority.Name))
-		}
+		args.Set("authentication-config", cpath(kasVolumeAuthConfig().Name, AuthenticationConfigKey))
```
Comment (deads2k) [inline control-plane-operator/controllers/hostedcontrolplane/kas/config.go:177]: this should still be conditional. it's reasonable to having an intermediate state with no authentication configured and the kube-apiserver must be able to tolerate that.  SA tokens, client certificates, etc would still function.

**Comment by deads2k** [comment]
Trying to figure out how to get e2es since https://github.com/openshift/kubernetes/pull/1881 has merged

/test all

**Comment by deads2k** [comment]
> can't use `AudienceMatchPolicy` until upstream merges [kubernetes/kubernetes#123165](https://github.com/kubernetes/kubernetes/pull/123165) or we redirect our `k8s.io/apiserver` dep to the openshift fork

We'll need to copy/paste the types.

**Comment by deads2k** [comment]
config looks properly constructed to me, featuregate looks properly specified.

/lgtm

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 3]
```diff
@@ -2660,6 +2660,19 @@ func (r *HostedControlPlaneReconciler) reconcileKubeAPIServer(ctx context.Contex
 		return fmt.Errorf("failed to reconcile api server config: %w", err)
 	}
 
+	if !util.HCPOAuthEnabled(hcp) &&
+		len(hcp.Spec.Configuration.Authentication.OIDCProviders) > 0 {
+		authConfig := manifests.AuthConfig(hcp.Namespace)
+		if _, err := createOrUpdate(ctx, r, authConfig, func() error {
+			return kas.ReconcileAuthConfig(authConfig,
+				p.OwnerRef,
+				p.ConfigParams(),
```
Comment (stlaz) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:2669]: Are these config params already validated? E.g. you can reach all of the issuers from the OIDCProviders and their discovery returns valid data (per https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata), CA can verify the serving cert at the issuer URL, ...?

Can you point me to the validation code?

Comment (deads2k) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:2669]: They weren't validated before either, right?  We probably validate in self-managed, but most of the guide rails are missing from the re-implementation.

**Pull request merged**